subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/cpu.h

```diff
@@ -16,12 +16,21 @@ static inline void cpu_sleep(void)
 #endif
 }
 
-static inline void cpu_dsb(void)
+static inline void cpu_dmb(void)
 {
-#if defined(CONFIG_CPU_CORTEX_M0) || defined(CONFIG_ARCH_POSIX)
-	/* No need of data synchronization barrier */
-#elif defined(CONFIG_CPU_CORTEX_M4) || defined(CONFIG_CPU_CORTEX_M33)
-	__DSB();
+#if defined(CONFIG_CPU_CORTEX_M0) || \
+	defined(CONFIG_CPU_CORTEX_M4) || \
+	defined(CONFIG_CPU_CORTEX_M33)
+	/* NOTE: __DMB() shall have a compiler memory barrier (clobber)
+	 *      aka __asm__ volatile ("dmb sy" : : : "memory");
+	 */
+	__DMB();
+#elif defined(CONFIG_ARCH_POSIX)
+	/* FIXME: Add necessary host machine required Data Memory Barrier
+	 *        instruction alongwith the below defined compiler memory
+	 *        barrier (clobber).
+	 */
+	__asm__ volatile ("" : : : "memory");
```

Comment:
Could this be applied in the case above, as well?

Reply:
This solves the issue, but following the ARM recommendation, will keep the DMB instruction.

Reply:
A reason not to do this is, if you ever want this code to switch to using arch_ APIs; eventually we're going to get an ARM or even a cross-ARCH API to insert compiler barriers and avoid instruction re-ordering. Then if you already do what's needed here, migrating to generic API calls will be easier and backwards-compatible.

Reply:
Kernel code uses DSB/ISB instructions for both M0 and M4 architectures, maybe i understand the control path incorrectly, but they are there definitely in the cpu_idle.S, swap_helper.S to name a few.

When there is a cross-ARCH API available, the change can be done to reuse.

Reply:
Yeap, DSB, ISBs _are_ used, for reasons explained inline (not related to compiler barrier placements)